### PR TITLE
Fix submit failed after reload

### LINKED ISSUE #2964: Task suspension in ready state upon suite reload
Users have observed, in version ``7.8.1``, abnormalities with job submission timings whereby tasks spend longer than expected, or indeed appear to get stuck, in the ``ready`` state.

From further investigation we have deduced that it results from the reloading of a suite in the middle of job retries.

#### Minimal example

Content of ``suite.rc``:

```ini
[scheduling]
    [[dependencies]]
        graph = """
                foo:submit-fail => stopper
                reload_suite => stopper
                """
[runtime]
    [[foo]]
        [[[job]]]
            batch system = at
            batch submit command template = sleep 5
            submission retry delays = 3*PT5S
    [[stopper]]
        script = sleep 30; cylc stop "${CYLC_SUITE_NAME}"
    [[reload_suite]]
        script = cylc reload "${CYLC_SUITE_NAME}"
```

where, in order to get the reload to occur between retries, it is necessary to bodge the logic of the reload command within ``Scheduler.command_reload_suite()``, as follows:

```diff
diff --git a/lib/cylc/scheduler.py b/lib/cylc/scheduler.py
index f965b8b1a..2342a14d2 100644
--- a/lib/cylc/scheduler.py
+++ b/lib/cylc/scheduler.py
@@ -960,6 +960,7 @@ conditions; see `cylc conditions`.
         LOG.info("Reloading the suite definition.")
         old_tasks = set(self.config.get_task_name_list())
         self.suite_db_mgr.checkpoint("reload-init")
+        sleep(15)
         self.load_suiterc(is_reload=True)
```

Observed behaviour is that the ``foo`` task remains in the ``ready`` state, never progressing to ``submit-failed`` as would be expected & as occurs when the ``reload_suite`` task is removed from the example suite definition.

### PR DESCRIPTION
**Pull request by matthewrmshin**
Fix #2964.

### CODE REVIEW

File: tests/jobscript/11-shell-ksh.t [hunk 1]
```diff
@@ -25,7 +25,7 @@ set_test_number 5
 install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
 run_ok "${TEST_NAME_BASE}-validate" \
     cylc validate "${SUITE_NAME}" --set=KSH="${KSH}"
-sed -in 's/^.* WARNING - //p' "${TEST_NAME_BASE}-validate.stderr"
+sed -in 's/^.*WARNING - //p' "${TEST_NAME_BASE}-validate.stderr"
```
Comment (matthewrmshin) [inline tests/jobscript/11-shell-ksh.t]: This change is not really related to the fix.

**Comment by matthewrmshin** [comment]
(This change definitely fixes the original problem on manual testing. I'm challenging myself with an automated test for this change at the moment - not succeeding yet.)

**Comment by kinow** [comment]
I registered a suite with the example from @sadielbartholomew and applied her patch to `scheduler.py`, but the suite was executed successfully running from `master` branch, and also from this pull request.

Had a look at the change in this pull request, and it appear to make sense to tell the scheduler it needs to process the tasks in case of a submit-fail. But without isolating a situation in the code that causes the issue in a deterministic way, can't think of a way of writing a unit test (even with mock), nor an automated test.

Only risk of merging as is, is that this issue can re-appear later on when we change task related objects or scheduler (or maybe due to some Py3 change?). 😢

**Comment by matthewrmshin** [comment]
Yes, it is quite hard to repeat the issue reliably - because it relies on jobs-submit command returning and reload being processed within a single iteration of the main loop (and in that order).

Yes, the problem may reappear if we are not careful with many of the upcoming changes with task job management (e.g. my job cluster support proposal), etc.

Can't see this being reintroduced by 2to3 changes - just 2 boolean flags here.

**Comment by matthewrmshin** [comment]
Added a (potentially placebo) test.

**Comment by matthewrmshin** [comment]
That's annoying. I've run your example suite (with this branch + the sleep) 50 times yesterday without observing the hang up. On the other hand, the hang up is repeatable on master (+ sleep).

**Comment by sadielbartholomew** [comment]
>I've run your example suite (with this branch + the sleep) 50 times yesterday without observing the hang up

@matthewrmshin how about I run through it tomorrow while you watch over, in case I am doing something silly? That could well be the case! :grinning:

**Comment by hjoliver** [comment]
(I'll wait on the result of your combined test, before I take a look).

**Comment by matthewrmshin** [comment]
OK. Got to the bottom of the issue. I was able to repeat @sadielbartholomew issue by fiddling with the sleep time. Unfortunately, the full fix is much bigger...

File: tests/jobscript/11-shell-ksh.t (hunk 1 above)
Comment (oliver-sanders) [reply tests/jobscript/11-shell-ksh.t]: Dammit, missed one!

**Comment by oliver-sanders** [comment]
Trying to get my head around the test, looks like it has the potential to be flaky.

Out of interest would it be possible to test this with a Python script by initilising a `Scheduler` object and patching a sleep onto the reload method? Not ideal but would remove the issue of timing from the test.

**Comment by matthewrmshin** [comment]
Test is anti-flaky - in actual fact - it passes most of the time even without the fix.

Yes, but not sure how to do this in a multi-test environment in the current test framework without making a lot of archaic adjustments.

**Comment by oliver-sanders** [comment]
> it passes most of the time even without the fix.

hmm, perhaps we would be better off without it?

**Comment by matthewrmshin** [comment]
Please do a Reddit style reviewer vote (with the emoji reaction button or otherwise). Do we want to have the anti-flaky test or not? :+1:  or :-1:  I'll delete the test if we have a negative trend.

**Comment by kinow** [comment]
Voted not, but would vote not for no test too. I'd prefer to try helping with a test (functional or unit, as long as it is easy to maintain, and stable)

**Comment by matthewrmshin** [comment]
Several considerations here:
* Unit test is not appropriate, because the main issue is on how reload logic interact with job submission callback logic, and the subsequent state update logic in the main loop.
* A real test is only possible is we inject logic to slow down reload in various way. We can do this obviously, but are we going to introduce some funny logic that may become difficult to maintain?
* Lastly, the real culprit here is the necessity, on reload, to create a replacement `TaskProxy` instance. If the original `TaskProxy` instance can be modified in-place, then we may have a winner. However, `TaskProxy.__init__` is quite complicated - so I don't really want to tackle that yet - as it may break other things. I believe we still want to get this out for 7.8.2.

**Comment by kinow** [comment]
Got it. Won't veto based on no-tests. But maybe thenwe can create a separate issue to work on this test.

I think this might be possible with Python, along the lines of what Oliver suggested before.

And if we have to, the test can use unittest.mock too.

**Comment by oliver-sanders** [comment]
Not pretty but we could do something along the lines of:

```python
import os
import sys
from time import sleep
from tempfile import mkdtemp

from cylc.scheduler import Scheduler
from cylc.scheduler_cli import main
from cylc.suite_srv_files_mgr import SuiteSrvFilesManager

tempdir = mkdtemp()
suiterc = os.path.join(tempdir, 'suite.rc')

with open(suiterc, 'w+') as suiterc_file:
    suiterc_file.write('''
        [scheduling]
            [[dependencies]]
                graph = foo
    ''')

reg = 'foo'
SuiteSrvFilesManager().register(reg, tempdir)

def sleep_decorator(fcn):
    def wrapper(*args, **kwargs):
        sleep(5)
        fcn(*args, **kwargs)
    return wrapper

Scheduler.command_reload_suite = sleep_decorator(
    Scheduler.command_reload_suite)

sys.argv = ['run', reg, '--host=localhost', '--no-detach']
scheduler = main(is_restart=False)
```

**Comment by matthewrmshin** [comment]
(What I had in mind is a lot uglier, but involves a lot less code to maintain. E.g. in where the original `sleep 15` insertion, we can do this instead:)

```python
try:
    sleep(float(os.getenv('CYLC_TEST_RELOAD_SLEEP')))
except (TypeError, ValueError):
    pass
```

**Comment by hjoliver** [comment]
> What I had in mind is a lot uglier, ..

You're right there :grin: !

**Comment by hjoliver** [comment]
What about monkey patching the lib/cylc source file in your test to add the sleep, instead of modifying the library for the test?

**Comment by matthewrmshin** [comment]
For all intends and purposes, I have updated the test to do exactly the same stuff as the original sample in #2964 - and added a :monkey: CLI to put the original amount of sleep in the right place. Unfortunately, I seem to have lost the ability to get the sample suite to fail on master within and outside of test battery in my work environment today.

It does not make any sense, hence, the test may be as anti-flaky as before but takes a bit longer to run. :sob:

**Comment by kinow** [comment]
I'm on `master`, with a very dummy suite that has:

```
[scheduling]
[[dependencies]]
graph="""
foo
"""

[runtime]
[[foo]]
script="blablabla"
[[[job]]]
execution retry delays = PT5S
```

It fails, and then in the `task_events_mgr.py` file, I have breakpoints in the `if/else` of `TaskEventsManager#_process_message_failed`. I can see the debugger stopping there twice. Once in the first time the task fails, then after a while when it is retried.

My understanding was that, for this bug to happen, in the second execution of `_process_message_failed`, `Scheduler#command_reload_suite` had to be executed.

I tried first calling `Scheduler#command_reload_suite()` during the second time `TaskEventsManager#_process_message_failed` was invoked (i.e. once my breakpoint in the first line of this method was reached, I navigated back in the stack frame to the frame with the `Scheduler` object, then evaluated the expression `self.command_reload_suite()`; which can be confirmed to be executed by looking at the console output, which contains the right logs for reloading a suite).

But never happened. Tried calling it every time the `_process_message_failed` was executed, but still nothing.

Added the `sleep(15)` just before `self.load_suiterc` in `Scheduler`, and tried again. The suite was still stalled correctly after the second execution.

With the debugger attached to the process, and jumping between frames, in theory we must be able to simulate exactly what we want for these tests - we can even include `sleep(??)` statements if necessary, with the advantage of the debugger allowing us to call the suite reload or retry whenever we want.

My guess is that finding the exact state for which this happens with the debugger, would be easier to capture the current stack variables & task pool state, and craft a deterministic test. But so far no luck 😢

**Comment by hjoliver** [comment]
Fine with me, given the evident difficulty of making a reliable test.  If you continue to have problems we can revisit this again later...
